Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Clickhouse/BppTransactionJoin.hs

```diff
@@ -0,0 +1,262 @@
+      rideFareAmount = "ride_fare_amount",
+      rideTripStartTime = "ride_trip_start_time",
+      rideStatus = "ride_status"
+    }
+
+type BppTransactionJoin = BppTransactionJoinT Identity
+
+instance ClickhouseValue RideStatus
+
+instance ClickhouseValue Common.BookingStatus
+
+instance ClickhouseValue Time.Months
+
+instance ClickhouseValue TripCategory
+
+$(TH.mkClickhouseInstances ''BppTransactionJoinT 'SELECT_FINAL_MODIFIER)
+
+findAllRideItems ::
+  CH.HasClickhouseEnv CH.APP_SERVICE_CLICKHOUSE m =>
+  Merchant ->
+  MerchantOperatingCity ->
+  Int ->
+  Int ->
+  Maybe Common.BookingStatus ->
+  Maybe (ShortId Ride) ->
+  Maybe DbHash ->
+  Maybe DbHash ->
+  UTCTime ->
+  UTCTime ->
+  UTCTime ->
+  m [QRE.RideItem]
+findAllRideItems merchant opCity limitVal offsetVal mbBookingStatus mbRideShortId mbCustomerPhoneDBHash mbDriverPhoneDBHash now from to = do
+  bppTransaction <-
+    CH.findAll $
+      CH.select $
+        CH.limit_ limitVal $
+          CH.offset_ offsetVal $
+            CH.filter_
+              ( \bppTransaction _ ->
+                  do
+                    bppTransaction.rideCreatedAt >=. from
+                    CH.&&. bppTransaction.rideCreatedAt <=. to
+                    CH.&&. bppTransaction.bookingProviderId CH.==. merchant.id
+                    CH.&&. (bppTransaction.bookingMerchantOperatingCityId CH.==. opCity.id)
+                    CH.&&. CH.whenJust_ mbRideShortId (\rsid -> bppTransaction.rideShortId CH.==. rsid)
+                    CH.&&. CH.whenJust_ mbCustomerPhoneDBHash (\cpdh -> bppTransaction.riderDetailsMobileNumberHash CH.==. (Text.pack . show . unDbHash) cpdh)
+                    CH.&&. CH.whenJust_ mbDriverPhoneDBHash (\dpdh -> bppTransaction.rideDetailsDriverNumberHash CH.==. Just ((Text.pack . show . unDbHash) dpdh))
+                    CH.&&. CH.whenJust_ mbBookingStatus (\bs -> mkBookingStatusVal bppTransaction CH.==. bs)
+              )
+              (CH.all_ @CH.APP_SERVICE_CLICKHOUSE bppTransactionJoinTTable)
+  return $ fmap mkRideItem bppTransaction
+  where
+    mkBookingStatusVal ride =
+      CH.if_ (ride.rideStatus CH.==.. CH.valColumn Ride.COMPLETED) (CH.valColumn Common.COMPLETED) $
+        CH.if_ ((ride.rideStatus CH.==.. CH.valColumn Ride.NEW) CH.&&.. (ride.rideCreatedAt CH.>.. CH.valColumn (addUTCTime (- (6 * 60 * 60) :: NominalDiffTime) now))) (CH.valColumn Common.UPCOMING) $
+          CH.if_ ((ride.rideStatus CH.==.. CH.valColumn Ride.NEW) CH.&&.. (ride.rideCreatedAt CH.<=.. CH.valColumn (addUTCTime (- (6 * 60 * 60) :: NominalDiffTime) now))) (CH.valColumn Common.UPCOMING_6HRS) $
+            CH.if_ ((ride.rideStatus CH.==.. CH.valColumn Ride.INPROGRESS) CH.&&.. (ride.rideTripStartTime CH.>.. CH.valColumn (Just (addUTCTime (- (6 * 60 * 60) :: NominalDiffTime) now)))) (CH.valColumn Common.ONGOING) $
+              CH.if_ (ride.rideStatus CH.==.. CH.valColumn Ride.CANCELLED) (CH.valColumn Common.CANCELLED) (CH.valColumn Common.ONGOING_6HRS)
```

Comment:
I have idea how to avoid these nested if_
will add here later, if it will work

Reply:
idea is next: instead of creating huge nested condition on db level, we can construct different small conditions on Haskell level, depending on `bookingStatus` param:

```
.................
                    CH.&&. CH.whenJust_ mbBookingStatus (\bs -> mkBookingStatusCond bs bppTransaction)
              )
              (CH.all_ @CH.APP_SERVICE_CLICKHOUSE bppTransactionJoinTTable)
  return $ fmap mkRideItem bppTransaction
  where
    subtract6Hrs = (addUTCTime (- (6 * 60 * 60) :: NominalDiffTime) now)

    mkBookingStatusCond Common.COMPLETED ride = ride.rideStatus CH.==. Ride.COMPLETED
    mkBookingStatusCond Common.UPCOMING ride = ride.rideStatus CH.==. Ride.NEW CH.&&. ride.rideCreatedAt CH.>. subtract6Hrs
    mkBookingStatusCond Common.UPCOMING_6HRS ride = ride.rideStatus CH.==. Ride.NEW CH.&&. ride.rideCreatedAt CH.<=. subtract6Hrs
    mkBookingStatusCond Common.ONGOING ride = ride.rideStatus CH.==. Ride.INPROGRESS CH.&&. ride.rideTripStartTime CH.>. Just subtract6Hrs
    mkBookingStatusCond Common.ONGOING_6HRS ride = ride.rideStatus CH.==. Ride.INPROGRESS CH.&&. ride.rideTripStartTime CH.<=. Just subtract6Hrs
    mkBookingStatusCond Common.CANCELLED ride = ride.rideStatus CH.==. Ride.CANCELLED
```

in this implementation extra operators `&&..` and others not required